x-pack/plugins/infra/server/graphql/nodes/resolvers.ts

```diff
@@ -32,20 +32,19 @@ export const createNodeResolvers = (
 } => ({
   InfraSource: {
     async map(source, args, { req }, info) {
-      const { groupBy, metrics, nodeType, nodesKey } = extractGroupByAndMetrics(info);
+      const { metrics, path } = extractPathsAndMetrics(info);
```

Comment:
There is a problem with extracting a single pair of `path` and `metrics` on the `map` level, which becomes apparent with a query like

```graphql
{
  source(id: "default") {
    id
    map(timerange: {interval: "1m", to: 1533721921000, from: 1533635537000}, filters: []) {
      nodesA: nodes(path: [{id: "1", type: hosts}]) {
        path {
          value
        }
        metricsA: metrics(metrics: [{type: count}]) {
          name
          value
        }
        metricsB: metrics(metrics: []) {
          name
          value
        }
      }
      nodesB: nodes(path: [{id: "1", type: terms, field: "metricset.module"}]) {
        path {
          value
        }
        metrics(metrics: [{type: count}]) {
          name
          value
        }
      }
    }
  }
}
```

Via graphql field aliasing there could be multiple `metrics` or even `nodes` fields on a query. Currently, for multiple `metrics`, the later metrics simply overwrite the earlier. For multiple `nodes`, it even generates an invalid query.

The problem of multiple `metrics` is masked by the fact that, as commented below, the `metrics` argument to the field seems not to be used at all. Instead `[{ type: 'count' }]` is always assumed.

I can see two options to make it work:

1. Refactor the functions like `extractPathsAndMetrics` and others in the lib to support multiple queries at once.
2. Move the resolver logic down the resolver hierarchy so the graphql server takes care of calling it multiple times with different arguments.

In order to avoid a complexity explosion in the helper functions, I think option 2 would be preferable. It could lead to independent requests to the database for each (path, metric) tuple, but I wouldn't expect that to cause performance problems. If it does, we could add batching for optimization after the fact.

Reply:
We decided to defer this work till the next release. It's being tracked via https://github.com/elastic/kibana/issues/21812